editor/frontend/src/TransformElementsTool.ts

```diff
@@ -3,12 +3,160 @@
+    if (undefined === graphic)
+      return;
+
+    return {id, placement, graphic: IModelApp.renderSystem.createGraphicOwner(graphic)};
+  }
+
+  private disposeOfGraphics(): void {
+    this.data.forEach((data) => {
+      data.graphic.disposeGraphic();
+    });
+
+    this.data.length = 0;
+  }
+
+  private async cancelPendingRequests(): Promise<void> {
+    const requests = new Array<string>();
+    for (const [_key, id] of this.pending)
+      requests.push(id);
+
+    this.pending.clear();
+    if (0 === requests.length)
+      return;
+
+    return IpcApp.callIpcHost("cancelElementGraphicsRequests", this.iModel.key, requests);
+  }
+
+  /** Call to request a RenderGraphic for the supplied element id.
+ * @see [[cleanupGraphics]] Must be called when the tool exits.
+ */
+  public async createSingleGraphic(id: Id64String): Promise<boolean> {
+    try {
+      const info = await this.createRequest(id);
+
+      if (undefined !== info?.id)
+        this.pending.delete(info.id);
+
+      if (undefined === info?.graphic)
+        return false;
+
+      this.data.push(info);
+      return true;
+    } catch {
+      return false;
+    }
+  }
+
+  /** Call to request RenderGraphics for the supplied element ids. Does not wait for results as
+   * generating graphics for a large number of elements can take time. Instead an array of [[RenderGraphicOwner]]
+   * is populated as requests are resolved and the current dynamics frame displays what is available.
+   * @see [[cleanupGraphics]] Must be called when the tool exits.
+   */
+  public createGraphics(elements: Id64Arg): void {
+    if (0 === Id64.sizeOf(elements))
+      return;
+
+    try {
+      for (const id of Id64.iterable(elements)) {
+        const promise = this.createRequest(id);
+
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        promise.then((info) => {
+          if (undefined !== info?.id)
+            this.pending.delete(info.id);
+
+          if (undefined !== info?.graphic)
+            this.data.push(info);
+        });
+      }
+    } catch { }
+  }
+
+  /** Call to dispose of [[RenderGraphic]] held by [[RenderGraphicOwner]] and cancel requests that are still pending.
+   * @note Must be called when the tool exits to avoid leaks of graphics memory or other webgl resources.
+   */
+  public async cleanupGraphics(): Promise<void> {
+    await this.cancelPendingRequests();
+    this.disposeOfGraphics();
+  }
+
+  public addSingleGraphic(graphic: RenderGraphic, transform: Transform, context: DynamicsContext): void {
+    const branch = new GraphicBranch(false);
+    branch.add(graphic);
+
+    const branchGraphic = context.createBranch(branch, transform);
+    context.addGraphic(branchGraphic);
+  }
+
+  public addGraphics(transform: Transform, context: DynamicsContext): void {
+    for (const data of this.data)
+      this.addSingleGraphic(data.graphic, transform, context);

```

Comment:
Considered putting all the graphics into a single branch?

Reply:
done.